### PR TITLE
fix for teams auth

### LINKED ISSUE #390: Login fails when deployed to teams
**Project**

Calendar Skill 

**Description**

When asking "what's my schedule like in teams, the oauth card appears but the Login button doesn't respond.

**To Reproduce**

@vassist "What's my schedule like"

**Expected behavior**

Just like the emulator, it should pop up a browser for me to log in

**Screenshots**

nothing

**Additional context**

### PR DESCRIPTION
**Pull request by lauren-mills**
## Description

Adds TeamsAuthenticationMiddleware which converts the signin card button type to OpenUrl in Teams. 

## Related Issue

resolves #390

### CODE REVIEW

File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/TeamsAuthenticationMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    public class TeamsAuthenticationMiddleware : IMiddleware
+    {
+        public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = new CancellationToken())
+        {
+            // hook up onSend pipeline
+            turnContext.OnSendActivities(async (ctx, activities, nextSend) =>
+            {
+                foreach (var activity in activities)
+                {
+                    if (activity.ChannelId != "msteams")
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/TeamsAuthenticationMiddleware.cs]: i'd use equals("msteams", StringComparison.InvariantCultureIgnoreCase) to be safe

File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/TeamsAuthenticationMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                    {
+                        continue;
+                    }
+
+                    if (!activity.Attachments.Any())
+                    {
+                        continue;
+                    }
+
+                    if (activity.Attachments[0].ContentType != "application/vnd.microsoft.card.signin")
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/TeamsAuthenticationMiddleware.cs]: same here

**Review by lzc850612 — CHANGES_REQUESTED** [review]

File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/TeamsAuthenticationMiddleware.cs (hunk 1 above)
Comment (lauren-mills) [reply solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/TeamsAuthenticationMiddleware.cs]: Changed this to use the Channels class from the SDK

File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/TeamsAuthenticationMiddleware.cs (hunk 2 above)
Comment (lauren-mills) [reply solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/TeamsAuthenticationMiddleware.cs]: Done

**Review by lzc850612 — APPROVED** [review]

**Comment by lauren-mills** [comment]
@gabog check out this PR for the teams middleware. We took it out of master, but it will help get everything working for now.

**Comment by gabog** [comment]
Thanks @lauren-mills, this middleware worked, will this be re-added to the master at some point? Or should I put it in my own source?

**Comment by lauren-mills** [comment]
You should add it to your own source. We decided it would be better for us to push for the channel to be updated and include the workaround rather than add channel specific logic to master.

**Comment by iMicknl** [comment]
Is this still necessary? If you add the login URL to the `validDomains` section of your Bot Manifest (for Teams), it should work as expected. No work around needed. [Read more](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/authentication/auth-oauth-card#getting-started-with-oauthcard-in-teams)

**Comment by lauren-mills** [comment]
@iMicknl I believe it is still necessary if you're trying to test your bot and are not ready to publish it to the gallery. Please correct me if I'm misunderstanding the Teams docs, but it seems like the manifest is supplied to Teams during publishing, and if you just enable the Teams channels in the Bot Channels tab in the Azure Portal there's no way to supply it.

**Comment by iMicknl** [comment]
@lauren-mills the way to test (and debug) for Teams is to [sideload your bot in Teams](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/apps/apps-upload), even when you are not ready to publish it to the gallery / tenant. Otherwise you can't make use of special Teams features like tabs, oauth, messaging extensions and file support. 

>For bots designed only for personal conversations between a bot and a single user, see Adding a bot for personal chat only for an alternate way to access for testing purposes. ([source](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/apps/apps-upload))

@billbliss is above option still recommended for testing personal conversations? In my opinion it is confusing for developers this way, since it won't support oauth cards and file support for example. Should the documentation be updated or is this still recommended?

**Comment by lauren-mills** [comment]
Thanks @iMicknl, this is good to know. It would be a good idea to include this information in the Channels tab for bots in Azure, because at the moment you can enable the channel and start testing without any of this information. Hence, the confusion when things don't work.

**Comment by iMicknl** [comment]
@lauren-mills good suggestion, I filed an issue. https://github.com/Microsoft/BotFramework-Services/issues/42

**Pull request merged**